### PR TITLE
Reddit & GitHub Integration 🏷 [icons coming soon]

### PR DESCRIPTION
**Pull request by dlipeles**
#### TODO
- [ ] Get Icons #1898 
- [ ] Create PublisherOps Accounts for Reddit & GitHub 
- [ ] Fork omniauth-reddit to Brave-intl

#### Features

Allows users to register a Reddit or GitHub channel

#### How To Test

1. Register a test app with Reddit.com
2. Set the reddit client id and reddit client secret variables
3. Attempt to add a reddit channel from the dashboard

#### Pull Request Checklist

- [ ] Adequate test coverage exists to prevent regressions -- [Guide to Testing](https://guides.rubyonrails.org/testing.html)
- [ ] XSS is mitigated -- [Guide to XSS Prevention](https://guides.rubyonrails.org/security.html#cross-site-scripting-xss)
- [ ] No raw SQL -- [Guide to SQL Injection Prevention](https://guides.rubyonrails.org/security.html#sql-injection)
- [ ] UI/UX is responsive -- [Guide to Responsive UI/UX](https://developers.google.com/web/fundamentals/design-and-ux/responsive/)
- [ ] Integrated Matomo for new UI elements -- [Guide to Matomo](https://developer.matomo.org/guides/integrate-introduction)
- [ ] Passes checklist for Progressive Web App -- [Guide to PWAs](https://developers.google.com/web/progressive-web-apps/checklist)

### CODE REVIEW

File: app/controllers/publishers/omniauth_callbacks_controller.rb [hunk 1]
```diff
@@ -182,17 +181,46 @@ def register_vimeo_channel
...
+      @channel.details = RedditChannelDetails.new(
+        name: reddit_auth_hash.info.name,
+        reddit_channel_id: reddit_auth_hash.uid,
+        auth_provider: reddit_auth_hash.provider,
+        thumbnail_url: reddit_auth_hash.extra.raw_info.icon_img,
+        channel_url: "https://www.reddit.com/user/#{reddit_auth_hash.info.name}",
+        nickname: reddit_auth_hash.info.name,
+      )
+
+      existing_channel = Channel.joins(:reddit_channel_details).where("reddit_channel_details.reddit_channel_id": reddit_auth_hash.uid).first
```
Comment (corymcdonald) [inline app/controllers/publishers/omniauth_callbacks_controller.rb]: Let's ensure that the channel is verified when were finding the existing channel

File: app/models/github_channel_details.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
+class GithubChannelDetails < BaseChannelDetails
+  has_paper_trail
+
+  REDDIT_PREFIX = "github#channel:".freeze
```
Comment (corymcdonald) [inline app/models/github_channel_details.rb]:
```suggestion
  GITHUB_PREFIX = "github#channel:".freeze
```

File: app/models/github_channel_details.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  validates :github_channel_id, presence: true
+  validates :thumbnail_url, presence: true
+  validates :name, presence: true
+  validates :channel_url, presence: true
+
+  def channel_identifier
+    "#{GITHUB_PREFIX}#{github_channel_id}"
+  end
+
+  def authorizer_name
```
Comment (corymcdonald) [inline app/models/github_channel_details.rb]: do you need `authorizer_email` as well? All the other classes have it

File: db/migrate/20190516214846_create_reddit_channel_details.rb [hunk 4]
```diff
@@ -0,0 +1,14 @@
+class CreateRedditChannelDetails < ActiveRecord::Migration[5.2]
+  def change
+    create_table :reddit_channel_details, id: :uuid, default: -> { "uuid_generate_v4()"}, force: :cascade do |t|
+      t.string   "reddit_channel_id"
```
Comment (corymcdonald) [inline db/migrate/20190516214846_create_reddit_channel_details.rb]: I think we should add an index to this so that we don't run into any bottlenecks

Comment (dlipeles) [reply db/migrate/20190516214846_create_reddit_channel_details.rb]: Good idea c414343

**Pull request merged**